lib/events.js

```diff
@@ -240,23 +243,25 @@ EventEmitter.prototype.removeListener =
 
 EventEmitter.prototype.removeAllListeners =
     function removeAllListeners(type) {
-      var key, listeners;
+      var listeners, events;
 
-      if (!this._events)
+      if (!(events = this._events))
         return this;
 
       // not listening for removeListener, no need to emit
-      if (!this._events.removeListener) {
+      if (!events.removeListener) {
         if (arguments.length === 0)
           this._events = {};
-        else if (this._events[type])
-          delete this._events[type];
+        else if (events[type])
+          delete events[type];
         return this;
       }
 
       // emit removeListener for all listeners on all events
       if (arguments.length === 0) {
-        for (key in this._events) {
+        var keys = Object.keys(events);
```

Comment:
> you know that Object.keys uses the enum cache as well, right?

It still creates a new array every time though.

What you say about `delete obj.key` and for..in is true for Crankshaft but not TurboFan.  I'm not sure what we should be optimizing for.  Crankshaft is still the default but it's effectively in maintenance mode.  TurboFan is going to replace it some day but that may still be far off.  Or it may be next month, I don't know.  :-(